pkg/promutil/test/multiprojects_metric_scences.go

```diff
@@ -0,0 +1,538 @@
+package main
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/hanfei1991/microcosm/pkg/promutil"
+	"github.com/hanfei1991/microcosm/pkg/tenant"
+	"github.com/pingcap/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func main() {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		log.L().Info("Start http listen on :8083")
+		http.ListenAndServe(":8083", nil)
+	}()
+
+	go func() {
+		defer wg.Done()
+		log.L().Info("Start scenarios simulator")
+		simulator(wg)
+	}()
+
+	wg.Wait()
+	log.L().Info("Exit scenarios metric test")
+}
+
+// simulator will simulate the usage of multi-projects/tasks prometheus metric,
+// Includes following scenarios:
+
+/// intra-framework test, SHOULD NOT have same metric name
+/// using different service name in K8s to distinguish metric
+// 1. one servermaster and one executor
+// 2. one servermaster and multi-executors
+
+/// intra-app metric test, SHOULD NOT have same metric
+// 3. one jobmaster and one worker of same job type in one executor
+
+/// multi-tasks metric isolation
+// 4. multi-jobmasters and multi-workers of same job type in one executor
+
+/// cross app metric isolation
+// 5. multi-jobmasters and multi-workers of different job type in one executor
+
+/// app and framework metric isolation
+// 6. one jobmaster in one executor, has same original metric name
+
+/// tenant/project metric isolation
+// 7. multi-jobmasters of same job type for different project
+
+func simulator(wg sync.WaitGroup) {
+	scenes := []func(wg sync.WaitGroup){
+		scenarios1_OneServerOneExecutor, scenarios2_OneServerMultiExecutor,
+		scenarios3_OneJobmasterOneWorker, scenarios4_OneJobmasterMultiWorker,
+		scenarios5_MultiJobmasterMultiWorker, scenarios6_OneJobmasterOneExecutor,
+		scenarios7_MultiJobmasterMultiProjects,
+	}
+
+	for _, scene := range scenes {
+		scene(wg)
+	}
+}
+
+func scenarios1_OneServerOneExecutor(wg sync.WaitGroup) {
+	log.L().Info("Start scenarios1_OneServerOneExecutor simulation...")
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		// log.L().Info("Start scenarios1_OneServerOneExecutor servermaster")
+		// we create new registry here to simulate running an isolation process or container
+		reg := promutil.NewRegistry()
+		http.Handle("/metric1", promutil.HTTPHandlerForMetricImpl(reg))
+
+		// one server
+		factory := promutil.NewFactory4FrameworkImpl(reg)
+		counter := factory.NewCounter(prometheus.CounterOpts{
+			Namespace: "dfe",
+			Subsystem: "servermaster",
+			Name:      "counter",
+
+			ConstLabels: prometheus.Labels{
+				"service": "svr1",
+			},
+		})
+		counter.Add(1)
+
+		for {
+			counter.Add(0.1)
+			time.Sleep(time.Second)
+		}
```

Comment:
Better to release the goroutine.

Reply:
This main test is used for continous simulation test.
User can use `CTRL+C` to terminate the test.
Seems ok to me.